sympy/discrete/transforms.py

```diff
@@ -0,0 +1,240 @@
+
+ifft.__doc__ = fft.__doc__
+
+
+#----------------------------------------------------------------------------#
+#                                                                            #
+#                         Number Theoretic Transform                         #
+#                                                                            #
+#----------------------------------------------------------------------------#
+
+def _number_theoretic_transform(seq, p, inverse=False):
+    """Utility function for the Number Theoretic transform (NTT)"""
+
+    if not iterable(seq):
+        raise TypeError("Expected a sequence of integer coefficients " +
+                        "for Number Theoretic Transform")
+
+    p = as_int(p)
+    if isprime(p) == False:
+        raise ValueError("Expected prime modulus for " +
+                        "Number Theoretic Transform")
+
+    a = [as_int(x) for x in seq]
+
+    n = len(a)
+    if n < 1:
+        return a
+
+    b = n.bit_length() - 1
+    if n&(n - 1):
+        b += 1
+        n = 2**b
+
+    if (p - 1) % n:
+        raise ValueError("Expected prime modulus of the form (m*2**k + 1)")
+
+    a += [0]*(n - len(a))
+    for i in range(1, n):
+        j = int(ibin(i, b, str=True)[::-1], 2)
+        if i < j:
+            a[i], a[j] = a[j] % p, a[i] % p
+
+    pr = primitive_root(p)
+
+    rt = pow(pr, (p - 1) // n, p)
+    if inverse:
+        rt = pow(rt, p - 2, p)
+
+    w = [1]*(n // 2)
+    for i in range(1, n // 2):
+        w[i] = w[i - 1] * rt % p
+
+    h = 2
+    while h <= n:
+        hf, ut = h // 2, n // h
+        for i in range(0, n, h):
+            for j in range(hf):
+                u, v = a[i + j], a[i + j + hf]*w[ut * j]
+                a[i + j], a[i + j + hf] = (u + v) % p, (u - v) % p
+        h *= 2
+
+    if inverse:
+        rv = pow(n, p - 2, p)
+        for i in range(n):
+            a[i] = a[i]*rv % p
+
+    return a
+
+
+def ntt(seq, p):
+    r"""
+    Performs the Number Theoretic Transform (NTT), which specializes the
+    Discrete Fourier Transform (DFT) over quotient ring Z/pZ for prime p
+    instead of complex numbers C.
+
+
+    The sequence is automatically padded to the right with zeros, as the
+    radix 2 NTT requires the number of sample points to be a power of 2.
+
+    Parameters
+    ==========
+
+    seq : iterable
+        The sequence on which DFT is to be applied.
+    p : Integer
+        Prime modulus of the form (m*2**k + 1) to be used for performing
+        NTT on the sequence.
+
+    Examples
+    ========
+
+    >>> from sympy import ntt, intt
+    >>> ntt([1, 2, 3, 4], 3*2**8 + 1)
+    [10, 643, 767, 122]
+    >>> intt(_, 3*2**8 + 1)
+    [1, 2, 3, 4]
+    >>> intt([1, 2, 3, 4], 3*2**8 + 1)
+    [387, 415, 384, 353]
+    >>> ntt(_, 3*2**8 + 1)
+    [1, 2, 3, 4]
+
+    References
+    ==========
+
+    .. [1] http://www.apfloat.org/ntt.html
+    .. [2] http://mathworld.wolfram.com/NumberTheoreticTransform.html
+
```

Comment:
A reference to Wikipedia could also be included: https://en.wikipedia.org/wiki/Discrete_Fourier_transform_(general)

Reply:
Okay, will do.